File: app/src/main/java/com/etesync/syncadapter/ui/CollectionMembersListFragment.java
package com.etesync.syncadapter.ui;

import android.accounts.Account;
import android.content.Context;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.ListFragment;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.etesync.syncadapter.AccountSettings;
import com.etesync.syncadapter.App;
import com.etesync.syncadapter.Constants;
import com.etesync.syncadapter.HttpClient;
import com.etesync.syncadapter.R;
import com.etesync.syncadapter.journalmanager.JournalManager;
import com.etesync.syncadapter.model.CollectionInfo;
import com.etesync.syncadapter.model.JournalEntity;
import com.etesync.syncadapter.model.JournalModel;

import java.util.List;

import io.requery.Persistable;
import io.requery.sql.EntityDataStore;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;

public class CollectionMembersListFragment extends ListFragment implements AdapterView.OnItemClickListener, Refreshable {
    private EntityDataStore<Persistable> data;
    private Account account;
    private CollectionInfo info;
    private JournalEntity journalEntity;
    private AsyncTask asyncTask;

    private TextView emptyTextView;

    public static CollectionMembersListFragment newInstance(Account account, CollectionInfo info) {
        CollectionMembersListFragment frag = new CollectionMembersListFragment();
        Bundle args = new Bundle(1);
        args.putParcelable(Constants.KEY_ACCOUNT, account);
        args.putSerializable(Constants.KEY_COLLECTION_INFO, info);
        frag.setArguments(args);
        return frag;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        data = ((App) getContext().getApplicationContext()).getData();
        account = getArguments().getParcelable(Constants.KEY_ACCOUNT);
        info = (CollectionInfo) getArguments().getSerializable(Constants.KEY_COLLECTION_INFO);
        journalEntity = JournalModel.Journal.fetch(data, info.getServiceEntity(data), info.uid);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.collection_members_list, container, false);

        //This is instead of setEmptyText() function because of Google bug
        //See: https://code.google.com/p/android/issues/detail?id=21742
        emptyTextView = (TextView) view.findViewById(android.R.id.empty);
        return view;
    }

    @Override
    public void refresh() {
        asyncTask = new JournalMembersFetch().execute();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        refresh();

        getListView().setOnItemClickListener(this);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (asyncTask != null)
            asyncTask.cancel(true);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        final JournalManager.Member member = (JournalManager.Member) getListAdapter().getItem(position);

        new AlertDialog.Builder(getActivity())
                .setIcon(R.drawable.ic_info_dark)
                .setTitle(R.string.collection_members_remove_title)
                .setMessage(getString(R.string.collection_members_remove, member.getUser()))
                .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        DialogFragment frag = RemoveMemberFragment.newInstance(account, info, member.getUser());
                        frag.show(getFragmentManager(), null);
                    }
                })
                .setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                    }
                }).show();
    }

    class MembersListAdapter extends ArrayAdapter<JournalManager.Member> {
        MembersListAdapter(Context context) {
            super(context, R.layout.collection_members_list_item);
        }

        @Override
        @NonNull
        public View getView(int position, View v, @NonNull ViewGroup parent) {
            if (v == null)
                v = LayoutInflater.from(getContext()).inflate(R.layout.collection_members_list_item, parent, false);

            JournalManager.Member member = getItem(position);

            TextView tv = (TextView) v.findViewById(R.id.title);
            tv.setText(member.getUser());
            return v;
        }
    }

    private class JournalMembersFetch extends AsyncTask<Void, Void, JournalMembersFetch.MembersResult> {
        @Override
        protected MembersResult doInBackground(Void... voids) {
            try {
                AccountSettings settings = new AccountSettings(getContext(), account);
                OkHttpClient httpClient = HttpClient.create(getContext(), settings);
                JournalManager journalsManager = new JournalManager(httpClient, HttpUrl.get(settings.getUri()));

                JournalManager.Journal journal = JournalManager.Journal.fakeWithUid(journalEntity.getUid());
                return new MembersResult(journalsManager.listMembers(journal), null);
            } catch (Exception e) {
                return new MembersResult(null, e);
            }
        }

        @Override
        protected void onPostExecute(MembersResult result) {
            if (result.throwable == null) {
                MembersListAdapter listAdapter = new MembersListAdapter(getContext());
                setListAdapter(listAdapter);

                listAdapter.addAll(result.members);

                emptyTextView.setText(R.string.collection_members_list_empty);
            } else {
                emptyTextView.setText(result.throwable.getLocalizedMessage());
            }
        }

        class MembersResult {
            final List<JournalManager.Member> members;
            final Throwable throwable;

            MembersResult(final List<JournalManager.Member> members, final Throwable throwable) {
                this.members = members;
                this.throwable = throwable;
            }
        }
    }
}

File: app/src/main/java/com/etesync/syncadapter/ui/CollectionMembersListFragment.kt
package com.etesync.syncadapter.ui

import android.accounts.Account
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.ListFragment
import com.etesync.syncadapter.*
import com.etesync.syncadapter.journalmanager.JournalManager
import com.etesync.syncadapter.model.CollectionInfo
import com.etesync.syncadapter.model.JournalEntity
import com.etesync.syncadapter.model.JournalModel
import io.requery.Persistable
import io.requery.sql.EntityDataStore
import okhttp3.HttpUrl
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import java.util.concurrent.Future

class CollectionMembersListFragment : ListFragment(), AdapterView.OnItemClickListener, Refreshable {
    private lateinit var data: EntityDataStore<Persistable>
    private lateinit var account: Account
    private lateinit var info: CollectionInfo
    private lateinit var journalEntity: JournalEntity
    private var members: List<JournalManager.Member>? = null
    private var asyncTask: Future<Unit>? = null

    private var emptyTextView: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        data = (context!!.applicationContext as App).data
        account = arguments!!.getParcelable(Constants.KEY_ACCOUNT)!!
        info = arguments!!.getSerializable(Constants.KEY_COLLECTION_INFO) as CollectionInfo
        journalEntity = JournalModel.Journal.fetch(data, info.getServiceEntity(data), info.uid)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.collection_members_list, container, false)

        //This is instead of setEmptyText() function because of Google bug
        //See: https://code.google.com/p/android/issues/detail?id=21742
        emptyTextView = view.findViewById<View>(android.R.id.empty) as TextView
        return view
    }

    override fun refresh() {
        asyncTask = doAsync {
            try {
                val settings = AccountSettings(context!!, account)
                val httpClient = HttpClient.Builder(context, settings).build().okHttpClient
                val journalsManager = JournalManager(httpClient, HttpUrl.get(settings.uri!!)!!)

                val journal = JournalManager.Journal.fakeWithUid(journalEntity.uid)
                members = journalsManager.listMembers(journal)

                uiThread {
                    setListAdapterMembers(members!!)
                }
            } catch (e: Exception) {
                uiThread {
                    emptyTextView!!.text = e.localizedMessage
                }
            }
        }
    }

    private fun setListAdapterMembers(members: List<JournalManager.Member>) {
        val context = context
        if (context != null) {
            val listAdapter = MembersListAdapter(context)
            setListAdapter(listAdapter)

            listAdapter.addAll(members)

            emptyTextView!!.setText(R.string.collection_members_list_empty)
        }
    }

    override fun onResume() {
        super.onResume()
        if (members != null) {
            setListAdapterMembers(members!!)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        refresh()

        listView.onItemClickListener = this
    }

    override fun onDestroyView() {
        super.onDestroyView()
        if (asyncTask != null)
            asyncTask!!.cancel(true)
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val member = listAdapter.getItem(position) as JournalManager.Member

        AlertDialog.Builder(activity!!)
                .setIcon(R.drawable.ic_info_dark)
                .setTitle(R.string.collection_members_remove_title)
                .setMessage(getString(R.string.collection_members_remove, member.user))
                .setPositiveButton(android.R.string.yes) { dialog, which ->
                    val frag = RemoveMemberFragment.newInstance(account, info, member.user!!)
                    frag.show(fragmentManager!!, null)
                }
                .setNegativeButton(android.R.string.no) { dialog, which -> }.show()
    }

    internal inner class MembersListAdapter(context: Context) : ArrayAdapter<JournalManager.Member>(context, R.layout.collection_members_list_item) {

        override fun getView(position: Int, _v: View?, parent: ViewGroup): View {
            var v = _v
            if (v == null)
                v = LayoutInflater.from(context).inflate(R.layout.collection_members_list_item, parent, false)

            val member = getItem(position)

            val tv = v!!.findViewById<View>(R.id.title) as TextView
            tv.text = member!!.user

            val readOnly = v.findViewById<View>(R.id.read_only)
            readOnly.visibility = if (member.readOnly) View.VISIBLE else View.GONE

            return v
        }
    }

    companion object {

        fun newInstance(account: Account, info: CollectionInfo): CollectionMembersListFragment {
            val frag = CollectionMembersListFragment()
            val args = Bundle(1)
            args.putParcelable(Constants.KEY_ACCOUNT, account)
            args.putSerializable(Constants.KEY_COLLECTION_INFO, info)
            frag.arguments = args
            return frag
        }
    }
}
